Initial alpha code

Blackbaud-AlexKingman opened the pull request:
Design doc:
https://blackbaudinc.app.box.com/s/ojl6m4iv7rhjsicji7vp40wdeef2jq4v

Additional requirements:
- Ability to "confirm on close" when there are unsaved changes in responsive mode.
- ~~Iterators in "responsive mode"~~ (decided not MVP for this release).
- Workspace header should only show in “responsive mode”.

"Back to list" button design:
![image](https://user-images.githubusercontent.com/18354182/60517216-47d61b00-9cad-11e9-9942-058491a1be60.png)

Blackbaud-AlexKingman committed:
initial commit

Blackbaud-SteveBrush commented on package.json:
```diff
@@ -1,31 +1,41 @@
-  "version": "1.0.0",
-  "description": "Single-page-application for skyux-split-view",
+  "name": "@skyux/split-view",
+  "version": "3.0.0-alpha",
```
Just confirming, this will be `3.0.0-alpha.0`?

Blackbaud-AlexKingman replied:
Yep, I'll fix that.

Blackbaud-AlexKingman replied:
Done.

Blackbaud-SteveBrush commented on src/app/index.html:
```diff
@@ -1,12 +1,5 @@
-</div>
+<ul>
+  <li>
+    <a href="#" skyAppLink="/visual/split-view">Split view</a>
```
`href="#"` is no longer needed.

Blackbaud-AlexKingman replied:
Done.

Blackbaud-SteveBrush commented on src/app/public/index.ts:
```diff
@@ -1 +1 @@
-export * from './src/library.module';
+export * from './modules';
```
Probably would be a good idea to list out every type you wish to export, explicitly.

Blackbaud-AlexKingman replied:
Done.

Blackbaud-SteveBrush commented on src/app/public/modules/split-view/split-view.component.html:
```diff
@@ -0,0 +1,57 @@
+<div
+  class="sky-split-view"
+>
+  <div *ngIf="listVisible"
```
`isListVisible`

Blackbaud-AlexKingman replied:
Done.

Blackbaud-SteveBrush commented on src/app/public/modules/split-view/split-view.component.html:
```diff
@@ -0,0 +1,57 @@
+        <sky-icon
+          icon="chevron-left"
+        ></sky-icon>
+        Back to list
```
Should use `skyLibResources` pipe.

Blackbaud-AlexKingman replied:
Done.

Blackbaud-SteveBrush commented on src/app/public/modules/split-view/types/split-view-message-type.ts:
```diff
@@ -0,0 +1,3 @@
+export enum SkySplitViewMessageType {
+  FocusFirstItemInWorkspace = 0
```
You might be able to get away with a simpler and more generic message type that is easier to remember, something like `FocusWorkspace`?

Blackbaud-AlexKingman replied:
Done.

Blackbaud-SteveBrush reviewed (commented):
A few more findings during a manual test:
- Left arrow key moves the divider to the right and the right arrow key moves the divider to the left.
- Should the up and down arrows do anything? They move the divider but that seems odd to me.
- On page load, the animation is visibly moving the workspace into position. Possible to not animate on page load?
- Scroll bars are always visible on both the list and workspace.

<img width="1015" alt="Screen Shot 2019-06-07 at 1 57 03 PM" src="https://user-images.githubusercontent.com/12497062/59124713-6bb47400-892e-11e9-8a1b-b8180ed0932e.png">

Blackbaud-AlexKingman committed:
PR feedback

Blackbaud-AlexKingman commented:
@Blackbaud-SteveBrush I've seen to your comments above, plus the following items:
> Left arrow key moves the divider to the right and the right arrow key moves the divider to the left.

Fixed.

> Should the up and down arrows do anything? They move the divider but that seems odd to me.

I believe so. The controlling mechanism behind the scenes is an input range slider. Up and Down should work just like Left and Right - breaking this functionality could go against user expectations of someone trying to access the page with a screen reader.

> On page load, the animation is visibly moving the workspace into position. Possible to not animate on page load?

Fixed.

> Scroll bars are always visible on both the list and workspace.

I'm unable to replicate this? Even across Chrome/Safari/Firefox. Maybe this is a moot issue, but I'll still have other reviewers check for this to be sure.

Blackbaud-AlexKingman committed:
added iterator buttons

Blackbaud-AlexKingman committed:
added confirm to close

Blackbaud-AlexKingman committed:
100% coverage

Blackbaud-AlexKingman committed:
added e2e test

Blackbaud-AlexKingman committed:
e2e tests

Blackbaud-AlexKingman committed:
style tweaks

Blackbaud-AlexKingman committed:
removed extra exports and index.ts files

Blackbaud-AlexKingman committed:
removed unnecessary variables

Blackbaud-AlexKingman committed:
fixed imports

Blackbaud-AlexKingman committed:
visual test / demo upgrade

Blackbaud-AlexKingman committed:
fixed package.json

Blackbaud-AlexKingman committed:
removed local screeshots

Blackbaud-AlexKingman committed:
updated gitignore

Blackbaud-AlexKingman commented:
@Blackbaud-ToddRoberts this is ready for you to look at.

Blackbaud-SteveBrush commented on src/app/public/index.ts:
```diff
@@ -1 +1,13 @@
-export * from './src/library.module';
+
+// Split view components/modules.
+export * from './modules/split-view/split-view-iterator.component';
```
Probably don't need to export the components, just the modules, services, and any types you want to be public.

Blackbaud-AlexKingman replied:
Done.

Blackbaud-SteveBrush commented on src/app/public/styles/_mixins.scss:
```diff
@@ -0,0 +1,4 @@
+@import "~@skyux/theme/scss/mixins";
+@import "~@skyux/theme/scss/_compat/mixins";
```
We can probably kill this directory.

Blackbaud-AlexKingman replied:
Done.

Blackbaud-SteveBrush commented on src/app/public/modules/split-view/split-view.module.ts:
```diff
@@ -0,0 +1,81 @@
+    BrowserAnimationsModule,
+    CommonModule,
+    FormsModule,
+    SkyConfirmModule,
```
This is being used by your visual tests; probably should be removed from the source module, and added to app-extras.

Blackbaud-AlexKingman replied:
Done.

Blackbaud-SteveBrush commented on src/app/public/modules/split-view/split-view.module.ts:
```diff
@@ -0,0 +1,81 @@
+  imports: [
+    BrowserAnimationsModule,
+    CommonModule,
+    FormsModule,
```
Importing the `FormsModule` feels off to me since the slider isn't really a form control. Is there another way to meet the functional requirements without putting `ngModel` on the input range? Angular Material doesn't use `ngModel` and it is accessible. https://material.angular.io/components/slider

Maybe there's a way to use the `input` without binding to the underlying `FormControl` so that we can avoid a peer dependency on `@angular/forms`.

Blackbaud-AlexKingman replied:
Great question. Let me take a look...

Blackbaud-AlexKingman replied:
Ok, I was able to remove it by leveraging the `(change)` event to set the width property. `FormsModule` dependency has been removed.

Blackbaud-SteveBrush commented on src/app/public/modules/split-view/split-view.component.ts:
```diff
@@ -0,0 +1,423 @@
+  public messageStream = new Subject<SkySplitViewMessage>();
+
+  @Input()
+  public ariaRole: string;
```
For the sake of documenting the user guidelines, what would be a realistic value for the `ariaRole`, here?

Blackbaud-AlexKingman replied:
@Blackbaud-MattGregg do you have any feedback on this? What would be a good `ariaRole` for a split view?

Blackbaud-MattGregg replied:
@Blackbaud-AlexKingman For the splitview, I think what we'd want is for each panel to be a region on the page to act as landmarks. This makes it easy for navigation. https://www.w3.org/TR/wai-aria-practices/examples/landmarks/region.html
Teams would need to be able to set a property for aria-label which would be the name for the region.

Blackbaud-AlexKingman replied:
@Blackbaud-MattGregg I can easily add the aria-labels to each panel. Thanks for the feedback! One last question - do we want to enforce the role of `region`, or is that something we just outline in the docs, and then provide an input for consumers to decide?

Blackbaud-AlexKingman replied:
In the meantime, I've updated roles & labels to be applied to each panel.

Blackbaud-SteveBrush replied:
Let's talk more about this offline.

Blackbaud-MattGregg replied:
Yes. I think each panel would be a region. These are the main sections of the page or form which would be relevant to have in the page outline and support someone navigating between them. Teams would need to put in aria-label values that make sense for their use and each needs to be unique (e.g., List of invoices, Invoice selected to approve)

Blackbaud-SteveBrush commented on src/app/public/modules/split-view/split-view.component.ts:
```diff
@@ -0,0 +1,423 @@
+
+  private xCoord = 0;
+
+  private _listVisible = true;
```
The order of the properties is a bit off.
https://docs.blackbaud.com/skyux-team-docs/style#class-properties

Blackbaud-AlexKingman replied:
Updated!

Blackbaud-SteveBrush commented on src/app/public/modules/split-view/split-view-media-query.service.ts:
```diff
@@ -0,0 +1,85 @@
+} from 'rxjs/Subscription';
+
+@Injectable()
+export class SkySplitViewMediaQueryService {
```
This is nearly a copy/paste from Flyout. Just a thought: could we put this file in a universal location and have both libraries use it?
https://github.com/blackbaud/skyux-flyout/blob/master/src/app/public/modules/flyout/flyout-media-query.service.ts

Blackbaud-AlexKingman replied:
I had the same thought... Should it be part of core's `SkyMediaQueryService`?

Blackbaud-SteveBrush replied:
We can chat about this in the morning, with the three of us.

Blackbaud-SteveBrush replied:
Let's hold off on any decisions for now.

Blackbaud-SteveBrush commented on src/app/public/modules/split-view/types/split-view-message-type.ts:
```diff
@@ -0,0 +1,7 @@
+export enum SkySplitViewMessageType {
+  FocusWorkspace = 0,
+  IteratorEnableNextButton = 1,
```
Think we could mimic the `SkyFlyoutMessageType` iterator names?
https://github.com/blackbaud/skyux-flyout/blob/master/src/app/public/modules/flyout/types/flyout-message-type.ts#L4-L7

Blackbaud-AlexKingman replied:
Oops, I meant to do that. Yes, good catch.

Blackbaud-AlexKingman replied:
Done.

Blackbaud-SteveBrush commented on src/app/public/modules/split-view/split-view.component.ts:
```diff
@@ -0,0 +1,423 @@
+  }
+
+  private applyAutofocus(): void {
+    const applyAutoFocus = this.coreAdapterService.applyAutoFocus(this.elementRef);
```
The name of the "core adapter" method `applyAutoFocus` is confusing because it's sharing the same name as this class's `applyAutoFocus` method (the core adapter method is returning a boolean, but this method is returning void). Initially, I thought `SkyCoreAdapterService.applyAutoFocus` meant you were applying focus to an element, but it's really checking if the element _has_ an input with an `autofocus` attribute. Probably too late to fix the method name (I see it was added recently), but something we might need to fix in a breaking change (something like `containsAutofocusElement` would make more sense).

Also, the `getFocusableChildrenAndApplyFocus` violates the "no use of the word 'and' in a method name". We should strive to only run one task per method.

Blackbaud-SteveBrush reviewed (commented):
Round #2. Looking good!

We definitely need @Blackbaud-ToddRoberts to look at this.

Blackbaud-AlexKingman committed:
pr feedback

Blackbaud-AlexKingman committed:
removed formsModule as a dependency

Blackbaud-AlexKingman committed:
updated aria roles/labels

Blackbaud-TrevorBurch committed:
Upgraded to Sky 3 and removed role inputs (#2)

* Upgraded to Sky 3 and removed role inputs

* Reverted peer dependencies that were funny due to the migration utility

Blackbaud-ToddRoberts reviewed (changes requested):
When in mobile view and using iterator buttons, focus is not being automatically set on the first input in the workspace like it is when clicking from the list.

Blackbaud-ToddRoberts commented on src/app/visual/split-view/split-view-visual.component.scss:
```diff
@@ -0,0 +1,33 @@
+  padding-left:10px;
+}
+
+::ng-deep .sky-split-view-workspace-content {
```
If this is going to be used for code demo, we should demonstrate the preferred approach of wrapping your content in a div and applying e.g. `class="sky-padding-even-large"`

Blackbaud-AlexKingman replied:
Done.

Blackbaud-AlexKingman committed:
fixed iterators not focusing

Blackbaud-AlexKingman commented:
> When in mobile view and using iterator buttons, focus is not being automatically set on the first input in the workspace like it is when clicking from the list.

@Blackbaud-ToddRoberts Great catch. This has been fixed.

Blackbaud-AlexKingman committed:
more test coverage for autofocus

Blackbaud-AlexKingman committed:
moved aria label inputs to child components

Blackbaud-AlexKingman committed:
removed duplicate exports

Blackbaud-SteveBrush commented on package.json:
```diff
@@ -1,31 +1,78 @@
+    "@angular/platform-browser-dynamic": "7.2.15",
+    "@angular/router": "7.2.15",
+    "@blackbaud/auth-client": "2.15.1",
+    "@pact-foundation/pact": "7.4.0",
```
Bumping this to `@pact-foundation/pact@8.2.6` will get rid of the vulnerability warning.

Blackbaud-AlexKingman replied:
Done.

Blackbaud-SteveBrush commented on skyuxconfig.json:
```diff
@@ -1,10 +1,26 @@
-  "appSettings": {
-    "myLibrary": {
-      "name": "My Library"
+  "skyuxModules": [
```
The `skyuxModules` property is no longer needed, since we're running Builder 3.

Blackbaud-AlexKingman replied:
Removed.

Blackbaud-SteveBrush commented on src/app/app-extras.module.ts:
```diff
@@ -3,18 +3,17 @@ import {
 
-// Specify entry components, module-level providers, etc. here.
+import {
+  AppSkyModule
```
We should remove the `AppSkyModule` since it's importing modules that the split view module should be providing.

Blackbaud-AlexKingman replied:
I'm confused. This was approved in a prior PR: https://github.com/blackbaud/skyux-split-view/pull/2
Has something changed since then?

Also, split-view shouldn't be providing things like repeater - that's only used for the demo. We could revert the state of this file back to where it was prior to that PR ☝️ ?

Blackbaud-SteveBrush replied:
Whoops, that definitely should have been caught (by me). Sorry about that. I've been removing that module in all other library migrations. For example: https://github.com/blackbaud/skyux-core/pull/84/files

Blackbaud-SteveBrush replied:
Yeah, for modules that are needed by the visual demo, we can just add those to the `exports` array.

Blackbaud-AlexKingman replied:
This has been updated.

Blackbaud-SteveBrush commented on src/app/public/index.ts:
```diff
@@ -1 +1,9 @@
-export * from './src/library.module';
+
+// Split view components/modules.
+export * from './modules/split-view/split-view-media-query.service';
```
Do we expect consumers to import and utilize the media query service?

Blackbaud-AlexKingman replied:
I initially thought so, but after talking with Trevor this is probably not needed. Removed.

Blackbaud-SteveBrush commented on src/app/public/modules/split-view/split-view-workspace.component.ts:
```diff
@@ -0,0 +1,42 @@
+
+@Component({
+  selector: 'sky-split-view-workspace',
+  template: `<ng-content></ng-content>`,
```
What's the benefit of putting the workspace and list logic in the split-view component, versus in their respective component classes? Seems like the list and workspace components should handle their own logic/template, if possible?

Blackbaud-AlexKingman replied:
Two major reasons:
1. These components are just shells for their respective `ng-content`.
2. The only logic performed on these components is updating their breakpoints. While that logic lives in each individual component (see `updateBreakpoint()), the parent split view component has to watch for changes to window resizing and consumers updating `listWidth`, and then call `updateBreakpoint()` on each child component to coordinate appropriate breakpoint changes to the DOM. At the time, a central authority on breakpoint updates seemed more clear than having that logic broken out an replicated across the two children components.

What do you think?

Blackbaud-SteveBrush replied:
That makes sense, thanks. At any rate, we can always change the behavior in the future, if needed.

Blackbaud-SteveBrush reviewed (commented):
A few more questions...

Blackbaud-AlexKingman committed:
bumped deps, updated skyuxconfig

Blackbaud-AlexKingman committed:
stopped exporting media service

Blackbaud-AlexKingman committed:
reverted pact-web update

Blackbaud-AlexKingman committed:
updated package-lock

Blackbaud-AlexKingman committed:
updated app-extras

Blackbaud-SteveBrush commented on src/app/public/modules/split-view/split-view-iterator.component.html:
```diff
@@ -0,0 +1,26 @@
+<div
+  class="sky-split-view-iterators"
+>
+  <button
```
Needs `type="button"`.

Blackbaud-SteveBrush commented on src/app/public/modules/split-view/split-view-iterator.component.html:
```diff
@@ -0,0 +1,26 @@
+    </sky-icon>
+  </button>
+  <button
+    class="sky-btn sky-btn-default"
```
Needs `type="button"`.

Blackbaud-SteveBrush commented on src/app/public/modules/split-view/split-view-iterator.component.html:
```diff
@@ -0,0 +1,26 @@
+  class="sky-split-view-iterators"
+>
+  <button
+    class="sky-btn sky-btn-default"
```
Looks like you might need to add some margin between the buttons?
![Screen Shot 2019-07-02 at 9 32 56 AM](https://user-images.githubusercontent.com/12497062/60516928-c088a780-9cac-11e9-8df0-60387ed62dda.png)

Blackbaud-SteveBrush commented on src/app/public/modules/split-view/split-view.module.ts:
```diff
@@ -0,0 +1,71 @@
+  ],
+  exports: [
+    SkySplitViewComponent,
+    SkySplitViewIteratorComponent,
```
You probably don't need to export the iterator component since it won't be used by the consumer.

Blackbaud-SteveBrush commented on src/app/public/modules/split-view/split-view.component.ts:
```diff
@@ -0,0 +1,414 @@
+    }
+  }
+
+  public get listWidth() {
```
Needs return type.

Blackbaud-AlexKingman committed:
Remove iterator and other things (#3)

* removed iterators, made back lable customizable

* renamed list to drawer

* renamed test

Blackbaud-SteveBrush commented on src/app/public/modules/split-view/split-view.component.ts:
```diff
@@ -0,0 +1,375 @@
+  public backLabel: string;
+
+  @Input()
+  public set drawerWidth(value: number) {
```
Remind me, we're going to try and put this input on the `drawer` component? If so, what do you think about naming it `defaultWidth` or `preferredWidth`?

Blackbaud-AlexKingman replied:
Yep, working on that. I've just got it named `width` for now, because a consumer could change it whenever they want and the template will react. The words `default` or `preferred` make me think there's something more to the intent, when all it really is doing is setting width. Would you be cool with just `width` ?

Blackbaud-SteveBrush replied:
The reason I requested a "looser" modifier is because the width can be changed by the user with the resize handle. So, the developer can set a "preferred width", but the user can override that width at will. Using a blatant "width" communicates that the width of the drawer is static. Is that accurate?

Blackbaud-AlexKingman replied:
All child-specific properties have been moved to the children components (width, ariaLabel). 👍

Blackbaud-SteveBrush commented on src/app/public/modules/split-view/split-view.component.ts:
```diff
@@ -0,0 +1,375 @@
+export class SkySplitViewComponent implements OnInit, AfterViewInit, OnDestroy {
+
+  @Input()
+  public backLabel: string;
```
What do you think about naming this `backButtonText`, to match our wording on other components' inputs? The `@skyux/modals` library is a good example; it uses "text" to describe the content of a button element in many places. See: https://github.com/blackbaud/skyux-modals/blob/master/src/app/public/modules/confirm/types/confirm-button-config.ts#L3

Blackbaud-AlexKingman replied:
Done.

Blackbaud-SteveBrush commented on src/app/public/modules/split-view/split-view.component.ts:
```diff
@@ -0,0 +1,375 @@
+  public messageStream = new Subject<SkySplitViewMessage>();
+
+  @Output()
+  public beforeWorkspaceClose = new EventEmitter<SkySplitViewBeforeWorkspaceCloseHandler>();
```
Do we still need this? It adds a lot of complexity just for the mobile view. Since we're assuming the developer will be handling all of the states and whether or not a form can be closed, the developer could focus the workspace with the message stream and launch an alert if the user opens the drawer and clicks on a link, if the form has unchanged work. Let me know if I'm missing a scenario.

Blackbaud-AlexKingman replied:
This is to prevent users from losing work when in responsive mode and clicking the "back to list" button.

Blackbaud-SteveBrush replied:
Correct. Can the drawer be shown overtop of the workspace, without removing the workspace from the DOM? If so, my description above should be a possible workaround.

Blackbaud-AlexKingman replied:
This has been removed.

Blackbaud-SteveBrush commented on src/app/public/modules/split-view/split-view.component.ts:
```diff
@@ -0,0 +1,375 @@
+  private updateBreakpoints(): void {
+    // Update drawer component.
+    if (this.drawerComponent) {
+      this.drawerComponent.updateBreakpoint(this.drawerWidth);
```
Since we already have access to the elements wrapping the drawer and workspace components, telling each component to resize itself feels like a scope breach. I would either:
1. Move the template and width logic to the drawer and workspace components. They would listen for breakpoint changes themselves and resize accordingly.
2. Remove the `updateBreakpoint` methods from the drawer and workspace components and set the width of the wrapping elements in the split-view component directly.

Blackbaud-AlexKingman replied:
This is also related to my attempt to put `width` on the drawer component. I'm already heading down this path, and I'll update this PR with that work soon.

Blackbaud-AlexKingman replied:
This has been restructured - The components no longer tell the children what to do - instead they just communicate with the `SplitViewService`.

Blackbaud-SteveBrush commented on src/app/public/modules/split-view/split-view.component.html:
```diff
@@ -0,0 +1,70 @@
+    (@workspaceEnter.done)="onWorkspaceEnterComplete($event)"
+  >
+
+    <div *ngIf="!isDrawerVisible"
```
Using `ngIf` removes the element from the DOM. Does this affect the accessibility of the drawer-workspace relationship? If so, maybe you could use `[hidden]` instead?

Blackbaud-AlexKingman replied:
That's a really good point. Hadn't thought of that. It might have some impact on those animations, because they rely on the element being destroyed. Let me look into it...

Blackbaud-SteveBrush replied:
Cool. FYI, this question relates to the `beforeWorkspaceClose` question, above.

Blackbaud-AlexKingman replied:
Done. Now using `[hidden]`.

Blackbaud-SteveBrush reviewed (commented):
A few things.

Blackbaud-AlexKingman committed:
changed input to backButtonText

Blackbaud-AlexKingman committed:
Break out child components (#4)

* broke out child components

* changed resize handle color

* removed unnecessary comment

* removed unnecessary method

* minor tweaks

* fixed label text

* drawer is hidden instead of destroyed when mobile

* created service

* moved hostlistner to workspace component

* minor tweaks

* fixed test

Blackbaud-AlexKingman committed:
Break out child components (#5)

* broke out child components

* changed resize handle color

* removed unnecessary comment

* removed unnecessary method

* minor tweaks

* fixed label text

* drawer is hidden instead of destroyed when mobile

* created service

* moved hostlistner to workspace component

* minor tweaks

* fixed test

* removed beforeClose and moved functionality to demo

* removed bad export

Blackbaud-AlexKingman committed:
fixed height issue when workspace is bigger than list

Blackbaud-AlexKingman committed:
changed popup text

Blackbaud-AlexKingman committed:
fixed maxwidth to go by parent instead of window

Blackbaud-TrevorBurch commented on e2e/split-view.e2e-spec.ts:
```diff
@@ -0,0 +1,42 @@
+    });
+  });
+
+  it('should match previous screenshot when drawer is displayed in mobile view (screen: xs)', (done) => {
```
So I understand why there is only an `xs` version of this test today; however, I wonder if we should put a note in a comment here reminding us why there isn't a `lg` version so that when we look at these tests in a few months we don't waste time remembering that.

Blackbaud-AlexKingman replied:
Done.

Blackbaud-TrevorBurch commented on skyuxconfig.json:
```diff
@@ -1,10 +1,23 @@
+      "browserSet": "speedy"
+    },
+    "unit": {
+      "browserSet": "paranoid"
```
I think these test settings are correct. However, I did notice in `core` that the `unit` setting is set to `speedy`. That feels like an issue in `core` to me. Do you agree?

Blackbaud-AlexKingman replied:
I have no idea. Let's ask @Blackbaud-SteveBrush about this when he gets back.

Blackbaud-TrevorBurch commented on src/app/public/index.ts:
```diff
@@ -1 +1,7 @@
-export * from './src/library.module';
+
```
Is there a reason for the extra blank line here?

Blackbaud-AlexKingman replied:
Nope. Removed.

Blackbaud-TrevorBurch commented on src/app/public/index.ts:
```diff
@@ -1 +1,7 @@
+
+// Split view types.
+export * from './modules/split-view/types/split-view-message';
+export * from './modules/split-view/types/split-view-message-type';
```
So, I understand the limiting of the use of barrels here. However, I did notice that we are note exporting the components here at all. Though they are not normally imported via Typescript by consumers - we have historically still exported them from the main barrel. This allows consumers to do things like `ViewChild` without finding the really long import path if needed. Seems odd that we would break that convention here.

Blackbaud-AlexKingman replied:
I don't have a lot of strong feelings on this one. However, it was a deliberate move to break convention. Steve and I talked about being more strict with our exports to avoid issues like cyclical dependencies. In a prior review ☝️ , we decided to remove any component/service/class/etc.. that we did not intend for consumers to use.

Blackbaud-TrevorBurch replied:
Ok. We can always add them in later.

Blackbaud-TrevorBurch commented on src/app/public/modules/split-view/fixtures/split-view.fixture.ts:
```diff
@@ -0,0 +1,57 @@
+  @ViewChild(SkySplitViewComponent)
+  public splitViewComponent: SkySplitViewComponent;
+
+  public hasUnsavedWork = false;
```
These final two public variables should be above the `ViewChild`

Blackbaud-AlexKingman replied:
Done.

Blackbaud-TrevorBurch replied:
I'm still seeing `splitViewMessageStream` and `width` below the view child.

Blackbaud-TrevorBurch commented on src/app/visual/split-view/split-view-visual.component.scss:
```diff
@@ -0,0 +1,29 @@
+
+  li {
+    padding: 10px 5px;
+    border-top:1px solid #ccc;
```
I think a few properties here are missing spaces after colons.

Blackbaud-AlexKingman replied:
Done.

Blackbaud-TrevorBurch commented on src/app/visual/split-view/split-view-visual.component.ts:
```diff
@@ -0,0 +1,91 @@
+    { id: 7, name: 'Jennifer Standley', amount: 4.53, date: '02/26/2019' }
+  ];
+
+  public set activeIndex(value: number) {
```
The getters and setters should be at the top of these properties.

Blackbaud-AlexKingman replied:
Done.

Blackbaud-TrevorBurch commented on src/app/visual/split-view/split-view-visual.component.ts:
```diff
@@ -0,0 +1,91 @@
+    this.activeIndex = 0;
+  }
+
+  public onItemClick(index: number) {
```
Return type

Blackbaud-AlexKingman replied:
Done.

Blackbaud-TrevorBurch commented on src/app/visual/split-view/split-view-with-repeater-visual.component.ts:
```diff
@@ -0,0 +1,95 @@
+import {
+  AfterViewInit,
+  Component,
```
`ChangeDetectorRef` should come second alphabetically.

Blackbaud-AlexKingman replied:
Done.

Blackbaud-TrevorBurch commented on src/app/visual/split-view/split-view-with-repeater-visual.component.ts:
```diff
@@ -0,0 +1,95 @@
+    { id: 7, name: 'Jennifer Standley', amount: 4.53, date: '02/26/2019' }
+  ];
+
+  public set activeIndex(value: number) {
```
Getters and setters should be first in the list of properties.

Blackbaud-AlexKingman replied:
Done.

Blackbaud-TrevorBurch commented on src/app/visual/split-view/split-view-with-repeater-visual.component.ts:
```diff
@@ -0,0 +1,95 @@
+  public ngAfterViewInit(): void {
+  }
+
+  public onItemClick(index: number) {
```
Return type

Blackbaud-AlexKingman replied:
Done.

Blackbaud-TrevorBurch commented on src/app/public/modules/split-view/split-view-drawer.component.scss:
```diff
@@ -0,0 +1,58 @@
+    display: none;
+  }
+
+  &::-ms-thumb {
```
I think we need to have styles for the `ms-track` too. I see the track portion when serving in IE

Blackbaud-AlexKingman replied:
Done. Nice catch!

Blackbaud-TrevorBurch commented on src/app/public/modules/split-view/split-view-drawer.component.ts:
```diff
@@ -0,0 +1,207 @@
+import {
+  AfterViewInit,
+  ChangeDetectionStrategy,
+  ChangeDetectorRef,
```
Alphabetically `Ref` goes before `Strategy`

Blackbaud-AlexKingman replied:
"Detection" vs "Detector"

Blackbaud-TrevorBurch commented on src/app/public/modules/split-view/split-view-drawer.component.ts:
```diff
@@ -0,0 +1,207 @@
+})
+export class SkySplitViewDrawerComponent implements AfterViewInit, OnInit, OnDestroy {
+
+  @Input()
```
We need to double check ordering here based on our style guide.

Blackbaud-AlexKingman replied:
Done.

Blackbaud-TrevorBurch replied:
Getters should go before normal public variables.

Blackbaud-TrevorBurch commented on src/app/public/modules/split-view/split-view-drawer.component.ts:
```diff
@@ -0,0 +1,207 @@
+  SkySplitViewService
+} from './split-view.service';
+
+let nextId = 0;
```
Since this will be global could we make it a little more specific?

Blackbaud-AlexKingman replied:
Done.

Blackbaud-TrevorBurch commented on src/app/public/modules/split-view/split-view-drawer.component.ts:
```diff
@@ -0,0 +1,207 @@
+    this.coreAdapterService.toggleIframePointerEvents(false);
+
+    Observable
+      .fromEvent(document, 'mousemove')
```
Do these work on mobile? Or do we need touch events too?

Blackbaud-AlexKingman replied:
Nope, but good thought. I just followed the flyout resize pattern here. Mind if we file this as a separate issue? It would take some considerable effort to test and this is just an alpha release.

Blackbaud-TrevorBurch replied:
Sure

Blackbaud-TrevorBurch commented on src/app/public/modules/split-view/split-view-media-query.service.ts:
```diff
@@ -0,0 +1,85 @@
+  SkyMediaQueryListener
+} from '@skyux/core';
+
+import {
```
Can we possibly combine the two `rxjs` imports into one?

Blackbaud-AlexKingman replied:
Done.

Blackbaud-TrevorBurch commented on src/app/public/modules/split-view/split-view-workspace-header.component.scss:
```diff
@@ -0,0 +1,14 @@
+@import "~@skyux/theme/scss/mixins";
+@import "~@skyux/theme/scss/_compat/mixins";
```
I thought that the main mixin import imported the compat one by default. Does it not?

Blackbaud-AlexKingman replied:
It does not. https://github.com/blackbaud/skyux-theme/blob/master/src/app/public/styles/_mixins-public.scss

Blackbaud-TrevorBurch replied:
Ok

Blackbaud-TrevorBurch commented on src/app/public/modules/split-view/split-view-workspace-header.component.ts:
```diff
@@ -0,0 +1,45 @@
+import {
+  SkySplitViewService
+} from './split-view.service';
+import { Subject } from 'rxjs';
```
Styling

Blackbaud-AlexKingman replied:
Done.

Blackbaud-TrevorBurch commented on src/app/public/modules/split-view/split-view-workspace.component.html:
```diff
@@ -0,0 +1,13 @@
+  role="region"
+  [attr.aria-label]="ariaLabel"
+>
+  <sky-split-view-workspace-header *ngIf="isMobile"
```
The spacing here seems odd. I think we would be good to either put this all on one line or the full open tag on one and the close on the next.

Blackbaud-AlexKingman replied:
Done.

Blackbaud-TrevorBurch commented on src/app/public/modules/split-view/split-view-workspace.component.html:
```diff
@@ -0,0 +1,13 @@
+  <div
+    class="sky-split-view-workspace-content sky-padding-even-large"
+  >
+  <ng-content></ng-content>
```
Missing a tab here.

Blackbaud-AlexKingman replied:
Done.

Blackbaud-TrevorBurch commented on src/app/public/modules/split-view/split-view-workspace.component.ts:
```diff
@@ -0,0 +1,99 @@
+import {
+  ChangeDetectionStrategy,
+  ChangeDetectorRef,
```
`Ref` before `Strategy`

Blackbaud-AlexKingman replied:
"Detection" vs "Detector"

Blackbaud-TrevorBurch commented on src/app/public/modules/split-view/split-view-workspace.component.ts:
```diff
@@ -0,0 +1,99 @@
+  ]
+})
+export class SkySplitViewWorkspaceComponent implements OnDestroy, OnInit {
+  public set isMobile(value: boolean) {
```
Return type. Also could we add a line before this?

Blackbaud-AlexKingman replied:
Setters don't have return types. I added a new line.

Blackbaud-TrevorBurch replied:
Oh. I forgot that it would yell at you for that. Ok.

Blackbaud-TrevorBurch commented on src/app/public/modules/split-view/split-view.component.scss:
```diff
@@ -0,0 +1,25 @@
+@import "~@skyux/theme/scss/mixins";
+@import "~@skyux/theme/scss/_compat/mixins";
```
Same question about the compat being included

Blackbaud-AlexKingman replied:
See answer above.

Blackbaud-TrevorBurch commented on src/app/public/modules/split-view/split-view.component.scss:
```diff
@@ -0,0 +1,25 @@
+  position: relative;
+  display: flex;
+  border: 1px solid $sky-border-color-neutral-medium;
+  background:#fff;
```
Missing space prior to `#fff`

Blackbaud-AlexKingman replied:
Done.

Blackbaud-TrevorBurch commented on src/app/public/modules/split-view/split-view.component.scss:
```diff
@@ -0,0 +1,25 @@
+  .sky-split-view-drawer-flex-container {
+    width: 100%;
+
+    @media (min-width: $sky-screen-sm-min) {
```
Shouldn't this be our new mixin?

Blackbaud-AlexKingman replied:
Yes, done.

Blackbaud-TrevorBurch commented on src/app/public/modules/split-view/split-view.component.spec.ts:
```diff
@@ -0,0 +1,520 @@
+  tick();
+}
+
+function getElementCords(elementRef: any) {
```
Return type

Blackbaud-AlexKingman replied:
Done.

Blackbaud-TrevorBurch commented on src/app/public/modules/split-view/split-view.component.spec.ts:
```diff
@@ -0,0 +1,520 @@
+  return document.querySelector('.sky-split-view-workspace-header-content') as HTMLElement;
+}
+
+function isWithin(actual: number, base: number, distance: number) {
```
Return type

Blackbaud-AlexKingman replied:
Done.

Blackbaud-TrevorBurch commented on src/app/public/modules/split-view/split-view.component.ts:
```diff
@@ -0,0 +1,188 @@
+  ]
+})
+export class SkySplitViewComponent implements OnInit, OnDestroy {
+  @Input()
```
Could we add a line above this?

Blackbaud-AlexKingman replied:
Done.

Blackbaud-TrevorBurch commented on src/app/public/modules/split-view/split-view.component.ts:
```diff
@@ -0,0 +1,188 @@
+
+  public previousButtonDisabled = false;
+
+  public get workspaceVisible() {
```
Return type

Blackbaud-AlexKingman replied:
Done.

Blackbaud-TrevorBurch commented on src/app/public/modules/split-view/split-view.component.ts:
```diff
@@ -0,0 +1,188 @@
+  @ContentChild(SkySplitViewDrawerComponent)
+  public drawerComponent: SkySplitViewDrawerComponent;
+
+  public set drawerVisible(value: boolean) {
```
Return type

Blackbaud-AlexKingman replied:
If you meant on the getter - done.

Blackbaud-TrevorBurch commented on src/app/public/modules/split-view/split-view.component.ts:
```diff
@@ -0,0 +1,188 @@
+  constructor(
+    private changeDetectorRef: ChangeDetectorRef,
+    private coreAdapterService: SkyCoreAdapterService,
+    private elementRef: ElementRef,
```
Personal preference - could we move `ElementRef` up one for alphabetical sorting?

Blackbaud-AlexKingman replied:
Wouldn't you think a developer would see the property name first? I feel like it takes less mental processing to look for the property name than it does to look for the name, but then also know the type. I really feel like alphabetizing the property name works better here.

Blackbaud-TrevorBurch replied:
Oh. You're right. Ok. That's fine

Blackbaud-TrevorBurch commented on src/app/public/modules/split-view/split-view.component.ts:
```diff
@@ -0,0 +1,188 @@
+    }
+  }
+
+  private handleIncomingMessages(message: SkySplitViewMessage) {
```
Return type

Blackbaud-AlexKingman replied:
Done.

Blackbaud-TrevorBurch commented on src/app/public/modules/split-view/split-view.module.ts:
```diff
@@ -0,0 +1,71 @@
+    BrowserAnimationsModule,
+    CommonModule,
+    SkySplitViewResourcesModule,
+    SkyIconModule
```
`Icon` before `Split`

Blackbaud-AlexKingman replied:
Done.

Blackbaud-TrevorBurch commented on src/app/public/modules/split-view/split-view.service.ts:
```diff
@@ -0,0 +1,100 @@
+import {
+  Injectable,
```
Alphabetical ordering

Blackbaud-AlexKingman replied:
Done.

Blackbaud-TrevorBurch commented on src/app/public/modules/split-view/split-view.service.ts:
```diff
@@ -0,0 +1,100 @@
+    this.drawerVisible.next(true);
+  }
+
+  public updateBackButtonText(text: string) {
```
Return type

Blackbaud-AlexKingman replied:
Done.

Blackbaud-TrevorBurch commented on src/app/public/modules/split-view/split-view.service.ts:
```diff
@@ -0,0 +1,100 @@
+    this._backButtonTextStream.next(text);
+  }
+
+  public updateDrawerWidth(drawerWidth: number) {
```
Return type

Blackbaud-AlexKingman replied:
Done.

Blackbaud-TrevorBurch reviewed (changes requested):
Got some comments for you. Mostly style

Blackbaud-AlexKingman committed:
PR feedback

Blackbaud-AlexKingman committed:
using responsive mixins

Blackbaud-TrevorBurch reviewed (changes requested):
Just a couple things

Blackbaud-AlexKingman committed:
property ordering
